### PR TITLE
Minor engine cleanups.

### PR DESCRIPTION
**Pull request by afck**
* In `AuthorityRound::on_prepare_block`, return immediately if not a
validator, avoid unnecessary `String` allocation and clone the
header only when needed.
* In `engines::validator_set::safe_contract` add some convenience
methods for contract calls, and use them in the old and new (POSDAO)
code.

### CODE REVIEW

File: ethcore/src/engines/authority_round/mod.rs [hunk 1]
```diff
@@ -1300,28 +1300,24 @@ impl Engine<EthereumMachine> for AuthorityRound {
 
 	/// Make calls to the randomness and validator set contracts.
 	fn on_prepare_block(&self, block: &ExecutedBlock) -> Result<Vec<SignedTransaction>, Error> {
-		// Genesis is never a new block, but might as well check.
-		let header = block.header().clone();
-		let first = header.number() == 0;
+		let opt_signer = self.signer.read();
+		let signer = match opt_signer.as_ref() {
+			Some(signer) => signer,
+			None => return Ok(Vec::new()), // We are not a validator, so we shouldn't call the contracts.
```
Comment (vkomenda) [inline ethcore/src/engines/authority_round/mod.rs]: Control is reordered here but it looks benign. Tests are needed however.

**Comment by varasev** [comment]
We need the same changes in the `afck-sim-malice` branch so that I could repeat the manual tests again before we merge this PR.

**Comment by afck** [comment]
OK, let's leave the more complex validator set cleanups out for now; I force-pushed the updated commit. (Sorry; I should have done them much earlier.)

File: ethcore/src/engines/authority_round/mod.rs (hunk 1 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: Not sure: We were returning if there was no signer before, too, except that we did it later; all that happened between those points was needlessly locking the client and creating a closure.
Do you have any particular additional tests in mind?

Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: Now a this function exits earlier for a non-validator node than before. Also now it doesn't make the "constant contract call" if our node is not a validator. I agree that that was probably a mistake in the old version. Maybe we could test that on the non-validator node in `posdao-test-setup` and check that logs don't contain anything suspicious?

Comment (varasev) [reply ethcore/src/engines/authority_round/mod.rs]: I suggest not to make these changes if we are unsure what we do or to try opening the same PR for upstream and request a review from the Parity team.

Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: > Also now it doesn't make the "constant contract call" if our node is not a validator.

It never did: It only created the closure, but didn't call it.

> the same PR for upstream

Upstream doesn't have the `on_prepare_block` method at all, so it doesn't apply.

Anyway, it's not a big deal: Feel free to close this PR. We can always do cleanups later.

Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: > It never did: It only created the closure, but didn't call it.

That's true. I was over-complicating things. I'm OK with this PR.

**Comment by vkomenda** [comment]
@afck, would it make sense to rebase the meaningful commits from this PR on top of the current `aura-pos`?

**Comment by afck** [comment]
Yes, I think so. Happy to give it a try tomorrow.

**Comment by afck** [comment]
I rebased it. Please take another look.
These should be only minor changes that don't alter any logic.

**Comment by vkomenda** [comment]
It looks better now!

**Pull request merged**